File: src/backtest/strategy/InventoryMarketMaker.cpp
#include "backtest/strategy/InventoryMarketMaker.hpp"

#include <cmath>

namespace cmf::backtest
{

InventoryMarketMaker::InventoryMarketMaker(InventoryMarketMakerConfig config)
    : config_{config}
{
}

void InventoryMarketMaker::onBook(StrategyContext& context, const BookSnapshot& book)
{
    if (book.bids.empty() || book.asks.empty())
    {
        return;
    }

    const auto mid = (book.bids.front().price + book.asks.front().price) / 2.0;
    const auto inventorySkew = context.inventory() * config_.skewPerUnit;
    const auto desiredBid = mid - config_.halfSpread - inventorySkew;
    const auto desiredAsk = mid + config_.halfSpread - inventorySkew;
    const bool canBuy = context.inventory() + config_.orderQuantity <= config_.inventoryLimit;
    const bool canSell = context.inventory() - config_.orderQuantity >= -config_.inventoryLimit;

    if (shouldReplace(bidOrder_, bidPrice_, desiredBid, canBuy))
    {
        if (bidOrder_.has_value())
        {
            context.cancel(*bidOrder_);
            bidOrder_.reset();
        }
    }

    if (shouldReplace(askOrder_, askPrice_, desiredAsk, canSell))
    {
        if (askOrder_.has_value())
        {
            context.cancel(*askOrder_);
            askOrder_.reset();
        }
    }

    if (canBuy && !bidOrder_.has_value())
    {
        bidOrder_ = context.placeLimit(Side::Buy, desiredBid, config_.orderQuantity);
        bidPrice_ = desiredBid;
    }

    if (canSell && !askOrder_.has_value())
    {
        askOrder_ = context.placeLimit(Side::Sell, desiredAsk, config_.orderQuantity);
        askPrice_ = desiredAsk;
    }
}

void InventoryMarketMaker::onFill(StrategyContext& context, const Fill& fill)
{
    (void)context;
    clearFilledOrder(fill);
}

bool InventoryMarketMaker::shouldReplace(const std::optional<OrderId> orderId, const Price currentPrice, const Price desiredPrice, const bool canQuote) const
{
    if (!orderId.has_value())
    {
        return false;
    }
    if (!canQuote)
    {
        return true;
    }
    return std::abs(currentPrice - desiredPrice) > config_.requoteThreshold;
}

void InventoryMarketMaker::clearFilledOrder(const Fill& fill)
{
    if (fill.remainingQuantity > 0.0)
    {
        return;
    }
    if (bidOrder_.has_value() && *bidOrder_ == fill.orderId)
    {
        bidOrder_.reset();
    }
    if (askOrder_.has_value() && *askOrder_ == fill.orderId)
    {
        askOrder_.reset();
    }
}

} // namespace cmf::backtest

File: strategies/inventory_market_maker.py
import cmf_backtester as bt

from strategy_app import BacktestStrategyApp


class InventoryMarketMaker:
    def __init__(
        self,
        order_quantity=1000.0,
        inventory_limit=5000.0,
        half_spread=0.00000005,
        requote_threshold=0.000000005,
        skew_per_unit=0.0,
    ):
        self.order_quantity = order_quantity
        self.inventory_limit = inventory_limit
        self.half_spread = half_spread
        self.requote_threshold = requote_threshold
        self.skew_per_unit = skew_per_unit
        self.bid_order_id = None
        self.ask_order_id = None
        self.bid_price = 0.0
        self.ask_price = 0.0

    def on_book(self, ctx, book):
        if not book.bids or not book.asks:
            return

        mid = (book.bids[0].price + book.asks[0].price) / 2.0
        inventory_skew = ctx.inventory * self.skew_per_unit
        desired_bid = mid - self.half_spread - inventory_skew
        desired_ask = mid + self.half_spread - inventory_skew
        can_buy = ctx.inventory + self.order_quantity <= self.inventory_limit
        can_sell = ctx.inventory - self.order_quantity >= -self.inventory_limit

        if self._should_replace(self.bid_order_id, self.bid_price, desired_bid, can_buy):
            ctx.cancel(self.bid_order_id)
            self.bid_order_id = None
        if self._should_replace(self.ask_order_id, self.ask_price, desired_ask, can_sell):
            ctx.cancel(self.ask_order_id)
            self.ask_order_id = None

        if can_buy and self.bid_order_id is None:
            self.bid_order_id = ctx.place_limit(bt.Side.Buy, desired_bid, self.order_quantity)
            self.bid_price = desired_bid
        if can_sell and self.ask_order_id is None:
            self.ask_order_id = ctx.place_limit(bt.Side.Sell, desired_ask, self.order_quantity)
            self.ask_price = desired_ask

    def on_fill(self, ctx, fill):
        del ctx
        if fill.remaining_quantity > 0:
            return
        if self.bid_order_id == fill.order_id:
            self.bid_order_id = None
        if self.ask_order_id == fill.order_id:
            self.ask_order_id = None

    def _should_replace(self, order_id, current_price, desired_price, can_quote):
        if order_id is None:
            return False
        if not can_quote:
            return True
        return abs(current_price - desired_price) > self.requote_threshold


class InventoryMarketMakerApp(BacktestStrategyApp):
    strategy_name = "InventoryMarketMaker"
    default_strategy_params = {
        "order_quantity": 1000.0,
        "inventory_limit": 5000.0,
        "half_spread": 0.00000005,
        "requote_threshold": 0.000000005,
        "skew_per_unit": 0.0,
    }

    def create_strategy(self, strategy_params):
        return InventoryMarketMaker(**strategy_params)


if __name__ == "__main__":
    InventoryMarketMakerApp().run()
